submit-api/src/Cardano/TxSubmit/Tx.hs

```diff
@@ -1,68 +1,112 @@
 
--- | Write the response recieved when tx has been submitted.
-writeTxSubmitResponse
-  :: TxSubmitVar
-  -> (SubmitResult ApplyMempoolPayloadErr)
-  -> IO ()
-writeTxSubmitResponse tsv submitRes =
-  atomically $ putTMVar (txRespond tsv) submitRes
+-- | Submit a transaction to a local node.
+submitTx
+  :: forall mode block.
+     LocalNodeConnectInfo mode block
+  -> (Either (Tx Byron) (Tx Shelley))
+  -> IO (Either TxSubmitError TxId)
+submitTx connectInfo byronOrShelleyTx =
+  case (localNodeConsensusMode connectInfo, byronOrShelleyTx) of
+    (ByronMode{}, Left tx) -> do
+      result <- liftIO $ Api.submitTx connectInfo (TxForByronMode tx)
+      case result of
+        TxSubmitSuccess -> pure $ Right (getTxIdForTx tx)
+        TxSubmitFailureByronMode err ->
+          pure $ Left (TxSubmitByronError err)
 
--- | Submit a tx and wait for the response. This is done as a pair of atomic
--- operations, to allow the tx to be read in one operation, submmited and then
--- the response written as a second operation. Doing this as a single atmomic
--- operation would not work as the other end of the submit/response pair need
--- to be operated on independently.
-submitTx :: TxSubmitVar -> GenTx ByronBlock -> IO (Either TxSubmitError TxId)
-submitTx tsv tx =
-  case tx of
-    ByronTx txid _ -> do
-      atomically $ putTMVar (txSubmit tsv) tx
-      submitRes <- atomically (takeTMVar $ txRespond tsv)
-      case submitRes of
-        SubmitSuccess -> pure $ Right txid
-        SubmitFail r -> pure $ Left (TxSubmitFail r)
-    ByronDlg {} -> pure $ Left $ TxSubmitBadTx "Delegation"
-    ByronUpdateProposal {} -> pure $ Left $ TxSubmitBadTx "Proposal"
-    ByronUpdateVote {} -> pure $ Left $ TxSubmitBadTx "UpdateVote"
+    (ByronMode{}, Right{}) ->
+      pure $ Left $ TxSubmitEraMismatchError EraMismatch {
+                ledgerEraName = "Byron",
+                otherEraName  = "Shelley"
+              }
+
+    (ShelleyMode{}, Right tx) -> do
+      result <- liftIO $ Api.submitTx connectInfo (TxForShelleyMode tx)
+      case result of
+        TxSubmitSuccess -> pure $ Right (getTxIdForTx tx)
+        TxSubmitFailureShelleyMode err ->
+          pure $ Left (TxSubmitShelleyError err)
+
+    (ShelleyMode{}, Left{}) ->
+      pure $ Left $ TxSubmitEraMismatchError EraMismatch {
+                ledgerEraName = "Shelley",
+                otherEraName  = "Byron"
+              }
+
+    (CardanoMode{}, tx) -> do
+      result <- Api.submitTx connectInfo (TxForCardanoMode tx)
+      case result of
+        TxSubmitSuccess -> pure $ Right (either getTxIdForTx getTxIdForTx tx)
+        TxSubmitFailureCardanoMode (ApplyTxErrByron err) ->
+          pure $ Left (TxSubmitByronError err)
+        TxSubmitFailureCardanoMode (ApplyTxErrShelley err) ->
+          pure $ Left (TxSubmitShelleyError err)
+        TxSubmitFailureCardanoMode (ApplyTxErrWrongEra mismatch) ->
+          pure $ Left (TxSubmitEraMismatchError mismatch)
```

Comment:
Likewise, I suspect these calls to `pure` can be factored out (and a single `pure $` placed before the `case`), since every case has one.